build-and-deploy-docs/create_mkdocs_config.py

```diff
@@ -0,0 +1,342 @@
+    no_spaces = re.sub(r"\s", "-", text.strip())
+    no_specials = re.sub(r"[^\w_-]", "", no_spaces)
+    return no_specials.casefold()
+
+
+@dataclass
+class Page:
+    "A page to be rendered."
+    title: str
+    filename: str = ""
+    tokens: list[Token] = field(default_factory=list)
+
+    def get_filename(self):
+        "Get the associated filename."
+        if not self.filename:
+            return get_heading_anchor(self.title) + ".md"
+        return self.filename
+
+
+def split_readme(readme_file: Path,
+                 docs_dir: Path,
+                 pipeline_repo: str) -> List[Dict[str, Path]]:
+    """
+    Split the README file into individual markdown files.
+
+    Return a list of {<title>: <filename>} dictionaries suitable for a MkDocs
+    nav element.
+    """
+    # pylint: disable=too-many-locals
+    img_dir = docs_dir / 'img'
+
+    docs_dir.mkdir(exist_ok=True)
+    img_dir.mkdir(exist_ok=True)
+
+    # Parse the original markdown file into a stream of tokens
+    with readme_file.open(encoding="utf-8") as infile:
+        tokens = MarkdownIt("gfm-like").parse(infile.read())
+
+    # Break the monolithic page into multiple pages on H2s. Name the pages by
+    # the content of their headings.
+    pages = [Page("Home", filename="index.md"), ]
+
+    # Simultaneously build up a corrected set of anchor links. As we're
+    # splitting one file into multiple, an anchor link like `#authors` will
+    # need to include the filename, like `repo-details.md#authors`.
+    anchor_pages = {}
+
+    # MarkdownIt treats headers as three sequential tokens: heading_open,
+    # inline (which will have child tokens), and heading_close. Iterate through
+    # with itertools.pairwise so that we'll have access to heading_open and
+    # inline simultaneously.
+    for token, next_token in itertools.pairwise(
+            itertools.chain(tokens, [None, ])):
+        if token.type == "heading_open":
+            heading_content = next_token.content
+
+            if token.tag == "h2":
+                # We've moved on to a new page
+                pages.append(Page(heading_content))
+
+            # Associate this anchor with the current page
+            anchor = get_heading_anchor(heading_content)
+            assert anchor not in anchor_pages
+            anchor_pages[anchor] = pages[-1].get_filename()
+
+        pages[-1].tokens.append(token)
+
+    def sanitize_link(url):
+        """
+        Sanitize a link within the README.
+
+        There are five cases to be handled:
+        1. Links outside of the repository - leave unchanged
+        2. File links within the docs/ folder - rewrite
+        3. Images - copy into docs/ and rewrite
+        4. File links within the repository - redirect to the code browser
+        5. Anchor links - rewrite to reference the newly split pages
+        """
+        link = urlparse(url)
+
+        if link.scheme or link.netloc:
```

Comment:
Is there actually a risk here? `file://` will only expose files that are visible on the server anyway, and this logic specifically intended to prevent such links from triggering extra files to be copied into the documentation.

Reply:
We want to avoid arbitrary filesystem paths though, right? I may be missing context.

Reply:
Thanks for the context - you clarified that this is just to render a link that appears in the docs.